File: receipt_import.py
"""Импорт PDF-чека: извлечение текста и попытка сопоставить цены с каталогом."""
import re
import sys

from material_catalog import get_catalog_entry, get_billing_meta
from product_aliases import match_product_alias
from receipt_template_engine import detect_template, list_template_summaries, parse_with_template
from supplier_db import get_or_create_supplier, load_supplier_prices, save_supplier_prices
from supplier_prices_util import normalize_price_record, receipt_price_to_piece

_PDF_INSTALL_HINT = (
    "Для импорта PDF нужен пакет pypdf.\n\n"
    "В папке проекта выполните:\n"
    "  .venv\\Scripts\\pip install pypdf\n\n"
    "Или добавьте поставщика вручную: Поставщики → Добавить."
)


def _pdf_import_error_message():
    if getattr(sys, "frozen", False):
        return "Импорт PDF недоступен в этой сборке."
    return _PDF_INSTALL_HINT


def extract_pdf_text(path):
    errors = []

    try:
        from pypdf import PdfReader

        return _extract_with_reader(PdfReader, path), None
    except ImportError as e:
        errors.append(str(e))
    except Exception as e:
        errors.append(f"pypdf: {e}")

    try:
        from PyPDF2 import PdfReader

        return _extract_with_reader(PdfReader, path), None
    except ImportError:
        pass
    except Exception as e:
        errors.append(f"PyPDF2: {e}")

    try:
        import fitz

        doc = fitz.open(path)
        parts = [doc[i].get_text() or "" for i in range(len(doc))]
        doc.close()
        text = "\n".join(parts)
        if text.strip():
            return text, None
    except ImportError:
        pass
    except Exception as e:
        errors.append(f"pymupdf: {e}")

    if errors:
        return None, _pdf_import_error_message()
    return None, _pdf_import_error_message()


def _extract_with_reader(reader_cls, path):
    reader = reader_cls(path)
    parts = []
    for page in reader.pages:
        parts.append(page.extract_text() or "")
    return "\n".join(parts)


def _parse_price_token(s):
    s = s.replace(" ", "").replace(",", ".")
    m = re.search(r"(\d+(?:\.\d{1,2})?)", s)
    if not m:
        return None
    try:
        return float(m.group(1))
    except ValueError:
        return None


def parse_receipt_text(text):
    lines = [ln.strip() for ln in (text or "").splitlines() if ln.strip()]
    supplier_name = lines[0] if lines else "Поставщик"
    items = []

    for ln in lines[1:]:
        low = ln.lower()
        if any(w in low for w in ("итого", "total", "сумма", "ндс", "всего")):
            continue
        m = re.search(r"([\d\s]+[.,]\d{2})\s*(?:руб|₽|rur)?\s*$", ln, re.I)
        if not m:
            m = re.search(r"(\d+(?:[.,]\d{2})?)\s*(?:руб|₽)?\s*$", ln, re.I)
        if not m:
            continue
        price = _parse_price_token(m.group(1))
        if price is None or price <= 0:
            continue
        name_part = ln[: m.start()].strip(" -:\t")
        if len(name_part) < 2:
            continue
        items.append({"raw_name": name_part, "unit_price": price})

    return {"supplier_name": supplier_name, "items": items}


def _match_item_key(raw_name):
    hit = match_product_alias(raw_name)
    return hit["item_key"] if hit else None


def _match_row_from_alias(raw_name):
    """Полная строка для импорта: item_key + единицы из чека."""
    return match_product_alias(raw_name)


def _price_row_from_receipt(item_key, product_name, receipt_price, receipt_unit=None, units_per_pack=None):
    billing = get_billing_meta(item_key)
    ru = receipt_unit or billing.get("receipt_unit", "piece")
    upp = units_per_pack if units_per_pack is not None else billing.get("units_per_pack", 1)
    cost = receipt_price_to_piece(receipt_price, ru, upp)
    return normalize_price_record(
        {
            "product_name": product_name,
            "unit_price_cost": cost,
            "receipt_unit": "piece",
            "units_per_pack": upp,
        }
    )


def apply_parsed_receipt(parsed, merge=True, template_mapped=None):
    supplier = get_or_create_supplier(parsed["supplier_name"])
    existing = load_supplier_prices(supplier.id) if merge else {}

    matched = []
    unmatched = []

    rows = template_mapped if template_mapped is not None else parsed["items"]
    for row in rows:
        item_key = row.get("item_key") if template_mapped is not None else None
        alias = None
        if item_key is None:
            alias = _match_row_from_alias(row["raw_name"])
            if alias:
                item_key = alias["item_key"]
        if item_key is None:
            item_key = _match_item_key(row["raw_name"])
        if item_key:
            entry = get_catalog_entry(item_key)
            product_name = row.get("product_name") or (alias or {}).get("product_name") or row["raw_name"]
            receipt_price = row.get("unit_price_receipt", row.get("unit_price", 0))
            rec = _price_row_from_receipt(
                item_key,
                product_name,
                receipt_price,
                receipt_unit=row.get("receipt_unit") or (alias or {}).get("receipt_unit"),
                units_per_pack=row.get("units_per_pack") if row.get("units_per_pack") is not None else (alias or {}).get("units_per_pack"),
            )
            if merge and item_key in existing and not rec["product_name"]:
                rec["product_name"] = existing[item_key].get("product_name", "")
            existing[item_key] = rec
            matched.append(
                {
                    **row,
                    "item_key": item_key,
                    "catalog_name": entry["name"] if entry else item_key,
                    "unit_price_cost": rec["unit_price_cost"],
                    "unit_price_client": rec["unit_price_client"],
                }
            )
        else:
            unmatched.append(row)

    save_supplier_prices(supplier.id, existing)
    return {
        "supplier": supplier,
        "matched": matched,
        "unmatched": unmatched,
    }


def parse_receipt_text_smart(text):
    tpl, score = detect_template(text)
    if tpl:
        parsed = parse_with_template(text, tpl)
        return parsed, {
            "template_id": tpl.get("id"),
            "template_name": tpl.get("name"),
            "score": score,
        }
    return parse_receipt_text(text), None


def import_receipt_pdf(path, merge=True):
    text, err = extract_pdf_text(path)
    if err:
        return None, err
    parsed, tpl_info = parse_receipt_text_smart(text)
    if not parsed.get("items"):
        n_templates = len(list_template_summaries())
        hint = (
            f" Загружено шаблонов: {n_templates}. "
            "Добавьте JSON в папку receipt_templates/ под формат вашего чека."
            if n_templates == 0
            else " Шаблон не подошёл — уточните detect/mappings в JSON или добавьте новый."
        )
        return None, "Не удалось найти позиции с ценами в PDF." + hint

    template_mapped = parsed.get("mapped_items") if tpl_info else None
    result = apply_parsed_receipt(parsed, merge=merge, template_mapped=template_mapped)
    result["raw_text_preview"] = (text or "")[:2000]
    result["parser"] = tpl_info or {"mode": "generic"}
    return result, None

File: flutter_app/lib/domain/receipt_import.dart
import 'dart:convert';

import 'package:file_picker/file_picker.dart';

import '../data/supplier_repository.dart';
import 'cost_calculator.dart';
import 'material_catalog.dart';
import 'product_aliases.dart';
import 'receipt_templates.dart';

class ParsedReceiptItem {
  ParsedReceiptItem({
    required this.rawName,
    required this.unitPrice,
    this.itemKey,
    this.receiptUnit,
    this.unitsPerPack,
    this.unitPriceReceipt,
  });

  final String rawName;
  final double unitPrice;
  final String? itemKey;
  final String? receiptUnit;
  final double? unitsPerPack;
  final double? unitPriceReceipt;
}

class ParsedReceipt {
  ParsedReceipt({required this.supplierName, required this.items});
  final String supplierName;
  final List<ParsedReceiptItem> items;
}

class ReceiptApplyResult {
  ReceiptApplyResult({
    required this.supplierId,
    required this.supplierName,
    required this.matched,
    required this.unmatched,
  });
  final int supplierId;
  final String supplierName;
  final List<Map<String, dynamic>> matched;
  final List<ParsedReceiptItem> unmatched;
}

double? _parsePriceToken(String s) {
  final cleaned = s.replaceAll(' ', '').replaceAll(',', '.');
  final m = RegExp(r'(\d+(?:\.\d{1,2})?)').firstMatch(cleaned);
  if (m == null) return null;
  return double.tryParse(m.group(1)!);
}

ParsedReceipt parseReceiptText(String text) {
  final lines = text.split('\n').map((l) => l.trim()).where((l) => l.isNotEmpty).toList();
  final supplierName = lines.isNotEmpty ? lines.first : 'Поставщик';
  final items = <ParsedReceiptItem>[];
  for (final ln in lines.skip(1)) {
    final low = ln.toLowerCase();
    if (['итого', 'total', 'сумма', 'ндс', 'всего'].any(low.contains)) continue;
    var m = RegExp(r'([\d\s]+[.,]\d{2})\s*(?:руб|₽|rur)?\s*$', caseSensitive: false).firstMatch(ln);
    m ??= RegExp(r'(\d+(?:[.,]\d{2})?)\s*(?:руб|₽)?\s*$', caseSensitive: false).firstMatch(ln);
    if (m == null) continue;
    final price = _parsePriceToken(m.group(1)!);
    if (price == null || price <= 0) continue;
    final namePart = ln.substring(0, m.start).replaceAll(RegExp(r'[ -:\t]+$'), '');
    if (namePart.length < 2) continue;
    items.add(ParsedReceiptItem(rawName: namePart, unitPrice: price));
  }
  return ParsedReceipt(supplierName: supplierName, items: items);
}

double receiptPriceToPiece(double unitPrice, String receiptUnit, double unitsPerPack) {
  if (unitPrice <= 0) return 0;
  final ru = receiptUnit.toLowerCase();
  if (ru == 'pack' || ru == 'уп' || ru == 'упак') {
    final n = unitsPerPack <= 0 ? 1.0 : unitsPerPack;
    if (n > 1) return double.parse((unitPrice / n).toStringAsFixed(4));
  }
  return unitPrice;
}

ParsedReceipt parseSaturnV1(String text, ReceiptTemplate tpl) {
  final items = <ParsedReceiptItem>[];
  final skip = ((tpl.data['skip_articles'] as List?) ?? []).map((e) => e.toString()).toSet();
  final config = (tpl.data['article_config'] as Map?)?.cast<String, dynamic>() ?? {};
  for (final code in config.keys) {
    if (skip.contains(code)) continue;
    final cfg = (config[code] as Map).cast<String, dynamic>();
    final itemKey = cfg['item_key'] as String?;
    if (itemKey == null) continue;
    final m = RegExp(
      '-${RegExp.escape(code)}[\\s\\S]*?'
      r'(\d+)\s+\S+[\s\n]*(\d+(?:[.,]\d+)?)\s+(\d+(?:[.,]\d+)?)',
    ).firstMatch(text);
    if (m == null) continue;
    final receiptPrice = double.tryParse(m.group(2)!.replaceAll(',', '.'));
    if (receiptPrice == null || receiptPrice <= 0) continue;
    final ru = cfg['receipt_unit'] as String? ?? 'piece';
    final upp = (cfg['units_per_pack'] as num?)?.toDouble() ?? defaultUnitsPerPack(itemKey).toDouble();
    final piece = receiptPriceToPiece(receiptPrice, ru, upp);
    final entry = catalogByItemKey(itemKey);
    items.add(ParsedReceiptItem(
      rawName: entry?.displayName ?? 'Сат-$code',
      unitPrice: piece,
      itemKey: itemKey,
      receiptUnit: ru,
      unitsPerPack: upp,
      unitPriceReceipt: receiptPrice,
    ));
  }
  final name = (tpl.data['supplier'] as Map?)?['fixed_name'] as String? ?? 'Сатурн';
  return ParsedReceipt(supplierName: name, items: items);
}

Future<(ParsedReceipt parsed, Map<String, dynamic>? meta)> parseReceiptTextSmart(String text) async {
  final templates = await loadReceiptTemplates();
  final (tpl, score) = detectTemplate(text, templates);
  if (tpl != null && tpl.customParser == 'saturn_v1') {
    return (parseSaturnV1(text, tpl), {'template_id': tpl.id, 'template_name': tpl.name, 'score': score});
  }
  if (tpl != null) {
    final parsed = parseReceiptText(text);
    final fixed = (tpl.data['supplier'] as Map?)?['fixed_name'] as String?;
    return (
      ParsedReceipt(supplierName: fixed ?? parsed.supplierName, items: parsed.items),
      {'template_id': tpl.id, 'template_name': tpl.name, 'score': score},
    );
  }
  return (parseReceiptText(text), null);
}

AliasHit? matchReceiptRow(ParsedReceiptItem item) {
  if (item.itemKey != null) {
    return AliasHit(
      itemKey: item.itemKey!,
      receiptUnit: item.receiptUnit,
      unitsPerPack: item.unitsPerPack,
      productName: catalogByItemKey(item.itemKey!)?.displayName,
    );
  }
  return matchProductAlias(item.rawName);
}

Future<ReceiptApplyResult> applyParsedReceipt(
  SupplierRepository repo,
  ParsedReceipt parsed, {
  bool merge = true,
}) async {
  final supplier = await repo.getOrCreateSupplier(parsed.supplierName);
  final existing = await repo.listPrices(supplier.id!);
  final byKey = {for (final p in existing) p.itemKey: p};

  final matched = <Map<String, dynamic>>[];
  final unmatched = <ParsedReceiptItem>[];

  for (final row in parsed.items) {
    final alias = matchReceiptRow(row);
    if (alias == null) {
      unmatched.add(row);
      continue;
    }
    final entry = catalogByItemKey(alias.itemKey);
    final ru = alias.receiptUnit ?? row.receiptUnit ?? defaultReceiptUnit(alias.itemKey);
    final upp = alias.unitsPerPack ??
        row.unitsPerPack ??
        defaultUnitsPerPack(alias.itemKey).toDouble();
    final cost = row.itemKey != null
        ? row.unitPrice
        : receiptPriceToPiece(row.unitPriceReceipt ?? row.unitPrice, ru, upp);
    final client = cost > 0 ? double.parse((cost * 1.1).toStringAsFixed(2)) : 0.0;
    final price = SupplierPrice(
      itemKey: alias.itemKey,
      productName: alias.productName ?? entry?.displayName ?? row.rawName,
      unitPriceCost: cost,
      unitPriceClient: client,
      receiptUnit: 'piece',
      unitsPerPack: upp,
    );
    if (merge && byKey.containsKey(alias.itemKey) && price.productName.isEmpty) {
      price.productName = byKey[alias.itemKey]!.productName;
    }
    await repo.upsertPrice(supplier.id!, price);
    matched.add({
      'raw_name': row.rawName,
      'item_key': alias.itemKey,
      'catalog_name': entry?.displayName ?? alias.itemKey,
      'unit_price_cost': cost,
      'unit_price_client': client,
    });
  }

  return ReceiptApplyResult(
    supplierId: supplier.id!,
    supplierName: supplier.name,
    matched: matched,
    unmatched: unmatched,
  );
}

Future<String?> pickReceiptText() async {
  final files = await FilePicker.pickFiles(
    type: FileType.custom,
    allowedExtensions: const ['pdf', 'txt', 'text'],
    allowMultiple: false,
  );
  if (files.isEmpty) return null;
  final f = files.first;
  final name = f.name.toLowerCase();
  final bytes = await f.readAsBytes();
  if (name.endsWith('.pdf')) {
    final text = extractPdfTextRough(bytes);
    return text.trim().isEmpty ? null : text;
  }
  return utf8.decode(bytes, allowMalformed: true);
}
